certbot/certbot/_internal/cli/helpful.py

```diff
@@ -166,75 +161,100 @@ def _usage_string(self, plugins: Iterable[str], help_arg: Union[str, bool]) -> s
 
         return usage
 
-    def remove_config_file_domains_for_renewal(self, parsed_args: argparse.Namespace) -> None:
+    def remove_config_file_domains_for_renewal(self, config: NamespaceConfig) -> None:
         """Make "certbot renew" safe if domains are set in cli.ini."""
         # Works around https://github.com/certbot/certbot/issues/4096
-        if self.verb == "renew":
-            for source, flags in self.parser._source_to_settings.items(): # pylint: disable=protected-access
-                if source.startswith("config_file") and "domains" in flags:
-                    parsed_args.domains = _Default() if self.detect_defaults else []
+        if (config.argument_sources['domains'] == ArgumentSource.CONFIG_FILE and
+                self.verb == "renew"):
+            config.domains = []
+
+    def _build_sources_dict(self) -> Dict[str, ArgumentSource]:
+        # ConfigArgparse's get_source_to_settings_dict doesn't actually create
+        # default entries for each argument with a default value, omitting many
+        # args we'd otherwise care about. So in general, unless an argument was
+        # specified in a config file/environment variable/command line arg,
+        # consider it as having a "default" value
+        result = { action.dest: ArgumentSource.DEFAULT for action in self.actions }
```

Comment:
Just to check, this doesn't result in creating a renewal config file with every possible value specified in it, right?

Reply:
it really shouldn't since `set_by_user` returns False for `ArgumentSource.DEFAULT`, buuuut might be worth throwing in a test just to make sure.